### ISSUE TITLE
Arraybuffers are "Externalized" when sending, but are not "un-externalized"

### ISSUE THREAD

**Issue by bfishman** [node-webrtc/node-webrtc#262]
I can send an arraybuffer to a peer just fine - but If I attempt to send the same array buffer to another peer immediately afterward, I get the following error:

```
FATAL ERROR: v8::ArrayBuffer::Externalize ArrayBuffer already externalized
Aborted (core dumped)
```

To "fix" this I simply have to re-create a new ArrayBuffer and send that to each additional client. But it'd be nice if you could reuse the same ArrayBuffer. Does it need to be "un-externalized" before the send function completes?  Perhaps it needs to be 'rewound'?

**Comment by bfishman** [comment]
On further research, it may not be possible to "unexternalize" the ArrayBuffer once externalized.  I may just end up copying all arrayBuffers before re-transmitting them, but it seems pretty inefficient. Will close, but please advise if you're aware of a better approach.

**Comment by modeswitch** [comment]
This is definitely a bug.

Taking a closer look at the code, we create an instance of `rtc::Buffer`, which in turn does a `memcpy` (see [here](https://github.com/js-platform/libwebrtc/blob/7848af36d86a4bbe8569693a1169123ad67c8905/webrtc/base/buffer.h#L60)). So it should be fine without invoking `Externalize` ([here](https://github.com/js-platform/node-webrtc/blob/55704ccd15cfd735e25360e6790f8ea0ab89d04e/src/datachannel.cc#L262)) and `Neuter` ([here](https://github.com/js-platform/node-webrtc/blob/55704ccd15cfd735e25360e6790f8ea0ab89d04e/src/datachannel.cc#L277)).

**Comment by modeswitch** [comment]
Here's a quick patch: 

```
diff --git a/src/datachannel.cc b/src/datachannel.cc
index 08d922a..c0184fc 100644
--- a/src/datachannel.cc
+++ b/src/datachannel.cc
@@ -258,7 +258,6 @@ NAN_METHOD(DataChannel::Send) {
       arraybuffer = view->Buffer();
     }

-    v8::ArrayBuffer::Contents content = arraybuffer->Externalize();
     rtc::Buffer buffer(content.Data(), content.ByteLength());

 #else
@@ -272,10 +271,6 @@ NAN_METHOD(DataChannel::Send) {

     webrtc::DataBuffer data_buffer(buffer, true);
     self->_jingleDataChannel->Send(data_buffer);
-
-#if NODE_MINOR_VERSION >= 11 || NODE_MAJOR_VERSION > 0
-    arraybuffer->Neuter();
-#endif
   }

   TRACE_END;
```

**Comment by modeswitch** [comment]
@bfishman Can you add a test case that reproduces this problem?

**Comment by modeswitch** [comment]
I suspect that externalizing buffers in this way will also cause a memory leak.

**Comment by vmolsa** [comment]
After ArrayBuffer had been etxrenalized, it does no longer owns the memory block. The caller should take steps to free memory when it is no longer needed.

**Comment by bfishman** [comment]
@modeswitch certainly can, will create a PR tonight (~14 hours)

**Comment by vmolsa** [comment]
@modeswitch Add [ArrayBuffer](https://github.com/vmolsa/webrtc-native/blob/master/src/ArrayBuffer.h) from webrtc-native to fix this problem. 

[here](https://github.com/vmolsa/webrtc-native/blob/master/src/DataChannel.cc#L161) is example of how to use it

**Comment by markandrus** [comment]
@vmolsa I see—so you Externalize and save the Contents of the ArrayBuffer on a hidden property, and you reuse that until it should be freed?

**Comment by vmolsa** [comment]
Yes. Contents will be free'd when gc loses all references to that object.

**Comment by bfishman** [comment]
[Pull Request](https://github.com/js-platform/node-webrtc/pull/264) with currently failing test case submitted

**Comment by MarcDelp** [comment]
Hi there, got the same issue :
When trying to send an ArrayBuffer one time there is no problem, but when trying to send it again (to an other peer) the ArrayBuffer has a byteLength of 0.
We fixed it in our code by creating a new ArrayBuffer each time we need to send the message.

An interesting point to notice is that in browsers such as Chrome or Firefox, you don't need to do that and you can send the same ArrayBuffer without creating a new one each time you want to send a message.

To see our project : https://github.com/coast-team/netflux

**Comment by Globik** [comment]
As for node.js version 6 and higher one need in the datachannel.cc replace - 249 - 
rtc::Buffer  buffer(static_cast<char*>(content.Data()), content.ByteLength());
with
rtc::CopyOnWriteBuffer buffer(static_cast<char*>(content.Data()),content.ByteLength())
Oops, tags away... char*

**Comment by getkey** [comment]
This fix requires `CopyOnWriteBuffer` which was introduced with commit `13041cf11f1c2691bb3ddc7d28e74cae9333b685`. So it first appears in `webrtc` v51, whereas markandrus/build-webrtc [builds v50](https://github.com/markandrus/build-webrtc/blob/8ad587bdde711aa15cd7b10f8ed4d8cb3fbd009a/config.default.json#L9).

I am currently compiling v51, if it or any subsequent version works I'll submit a pull request on markandrus/build-webrtc and here with @Globik's fix.

Compiling takes a few hours on my old computer though so I won't have the results before tomorrow.

**Comment by getkey** [comment]
Wait. I realized you described how to update to v51, which is unrelated to this issue.

**Comment by getkey** [comment]
I tried replacing [calls to `Externalize`](https://github.com/js-platform/node-webrtc/blob/f5d13d137a08d93b6bec48c82ef792f5df5b08b2/src/datachannel.cc#L249) (and `Neuter`) with the following chunk on code based on @vmolsa's work.

```c++
v8::Isolate *isolate = v8::Isolate::GetCurrent();
rtc::CopyOnWriteBuffer buffer;

if (arraybuffer->IsExternal()) {
	//I will complete this later
} else {
	v8::ArrayBuffer::Contents content = arraybuffer->Externalize();
	buffer = rtc::CopyOnWriteBuffer(static_cast<char*>(content.Data()),content.ByteLength());

	v8::ArrayBuffer* underlying = *arraybuffer;
	underlying->SetHiddenValue(v8::String::NewFromUtf8(isolate, "node::ArrayBuffer"), v8::External::New(isolate, content.Data()));
}
```

The compilation fails with:
```
../src/datachannel.cc: In static member function 'static Nan::NAN_METHOD_RETURN_TYPE node_webrtc::DataChannel::Send(Nan::NAN_METHOD_ARGS_TYPE)':
../src/datachannel.cc:281:21: error: 'class v8::ArrayBuffer' has no member named 'SetHiddenValue'; did you mean 'IntegerValue'?
         underlying->SetHiddenValue(v8::String::NewFromUtf8(isolate, "node::ArrayBuffer"), v8::External::New(isolate, content.Data()));
                     ^~~~~~~~~~~~~~
```

But as you can see in the [inheritance diagram](http://bespin.cz/~ondras/html/classv8_1_1ArrayBuffer.html), `v8::ArrayBuffer` inherits from `v8::Object` which has a public method `SetHiddenValue`.

@markandrus @vmolsa Do you have any guidance on that?

**Comment by markandrus** [comment]
Should be fixed by #338—will aim to release this in v0.0.64
